lib/net.c

```diff
@@ -822,18 +822,54 @@ static int
 init_port(struct gatekeeper_if *iface, uint16_t port_id,
 	uint8_t *pnum_succ_ports)
 {
-	int ret = rte_eth_dev_configure(port_id, iface->num_rx_queues,
+	struct rte_eth_dev_info dev_info;
+	uint64_t configured_rss_hf;
+	int ret;
+
+	memset(&dev_info, 0, sizeof(dev_info));
+	rte_eth_dev_info_get(port_id, &dev_info);
+
+	/* Only use RSS hash functions this device can handle. */
+	configured_rss_hf = gatekeeper_port_conf.rx_adv_conf.rss_conf.rss_hf;
+	gatekeeper_port_conf.rx_adv_conf.rss_conf.rss_hf &=
+		dev_info.flow_type_rss_offloads;
+
+	/*
+	 * Check whether this device supports the configured RSS hashes.
+	 *
+	 * It seems common that devices do not exactly support the
+	 * hashes in the DPDK macros such as ETH_RSS_IP, so until we choose
+	 * set of minimum hash functions required (instead of ETH_RSS_IP
+	 * which is overkill), issue a warning in that case.
```

Comment:
It seems that at least in the NIC that we've been using, it does not support all of the hash functions in `ETH_RSS_IP`.

The ixgbe driver indicates that it supports these hash functions:

    #define IXGBE_RSS_OFFLOAD_ALL ( \
        ETH_RSS_IPV4 | \
        ETH_RSS_NONFRAG_IPV4_TCP | \
        ETH_RSS_NONFRAG_IPV4_UDP | \
        ETH_RSS_IPV6 | \ 
        ETH_RSS_NONFRAG_IPV6_TCP | \
        ETH_RSS_NONFRAG_IPV6_UDP | \
        ETH_RSS_IPV6_EX | \
        ETH_RSS_IPV6_TCP_EX | \
        ETH_RSS_IPV6_UDP_EX)

But `ETH_RSS_IP` contains some hash functions that are not in this list:

    #define ETH_RSS_IP ( \
        ETH_RSS_IPV4 | \
        ETH_RSS_FRAG_IPV4 | \
        ETH_RSS_NONFRAG_IPV4_OTHER | \
        ETH_RSS_IPV6 | \
        ETH_RSS_FRAG_IPV6 | \
        ETH_RSS_NONFRAG_IPV6_OTHER | \
        ETH_RSS_IPV6_EX)

Since we have been using Gatekeeper with ixgbe, I suspect we only need some subset of the hashes in `ETH_RSS_IP`.